### PR TITLE
Add support for JAX v 4.10

### LINKED ISSUE #73: [MLIR] Missing bufferization for `memref.tensor_store`
This op is generated in our parameter-shift gradient method (for some cases) and will fail bufferization.

Fixed upstream last month in https://reviews.llvm.org/D144080, so a JAX update on our side should fix the issue.

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** This branch build on top of https://github.com/PennyLaneAI/catalyst/tree/jax-v0.4.10-upgrade

**Description of the Change:** Frontend changes to support JAX v0.4.10

**Benefits:** JAX v0.4.10 support

Closes #73 
[sc-38731]

### CODE REVIEW

File: .github/workflows/check-catalyst.yaml [hunk 1]
```diff
@@ -132,7 +132,7 @@ jobs:
     - name: Install Deps
       if: steps.cache-llvm-build.outputs.cache-hit != 'true'
       run: |
-        sudo apt-get install -y python3 python3-pip cmake ninja-build clang lld
+        sudo apt-get install -y python3 python3-pip cmake ninja-build clang lld ccache
```
Comment (dime10) [inline .github/workflows/check-catalyst.yaml:135]: Why are we adding ccache for llvm and mhlo? We only backup the cache files for the quantum dialect build so I think it would be useless.

File: .github/workflows/set_dep_versions.py [hunk 2]
```diff
@@ -21,19 +21,19 @@
 
 import requests
 
-jax_version = sys.argv[1]
+jax_version = "0.4.10"  # sys.argv[1]
```
Comment (dime10) [inline .github/workflows/set_dep_versions.py:24]:
```suggestion
jax_version = sys.argv[1]
```

File: frontend/catalyst/compilation_pipelines.py [hunk 3]
```diff
@@ -531,7 +531,7 @@ def _maybe_promote(self, function, *args):
         if needs_compile:
             if self.user_typed:
                 msg = "Provided arguments did not match declared signature, recompiling..."
-                warnings.warn(msg, UserWarning)
+                warnings.warn(msg, ResourceWarning)
```
Comment (dime10) [inline frontend/catalyst/compilation_pipelines.py:534]: Just curious why `ResourceWarning`, is there a leaked resource?

File: frontend/catalyst/jax_primitives.py [hunk 4]
```diff
@@ -20,9 +20,9 @@
 
 import jax
 import numpy as np
-from jax._src import util
+from jax._src import api_util, core, source_info_util, util
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py:23]: So from what I understand in the JAX repo everything is defined in `jax._src`, and a lot (but not all!) names are re-exported to the top-level as "public". 
For example, `jax.core.AbstractValue` in https://github.com/google/jax/blob/jaxlib-v0.4.10/jax/core.py.

Ideally, we would only access functions and classes from this public interface, but at the moment it looks like we can't get around importing some "hidden" ones.
But I think we should keep the number of those as small as possible and be explicit when we do so, whereas the changes in this file blanket import `core` from `jax._src` for example even though most of the classes we use are defined the public `jax.core`.

File: mlir/Makefile [hunk 5]
```diff
@@ -48,6 +48,8 @@ llvm:
 
 .PHONY: mhlo
 mhlo:
+	patch -d mlir-hlo/mhlo/analysis --follow-symlinks < patches/add-lmhlo-dependency.patch
+
```
Comment (dime10) [inline mlir/Makefile:52]: Oh wow, two new patches 😅 Would you mind explaining what they do and why?

File: mlir/lib/Gradient/IR/GradientOps.cpp [hunk 6]
```diff
@@ -28,6 +28,25 @@
...
+
+void JVPOp::setCalleeFromCallable(CallInterfaceCallable callee)
+{
+    (*this)->setAttr("callee", callee.get<SymbolRefAttr>());
+};
+
+void VJPOp::setCalleeFromCallable(CallInterfaceCallable callee)
+{
+    (*this)->setAttr("callee", callee.get<SymbolRefAttr>());
+};
```
Comment (dime10) [inline mlir/lib/Gradient/IR/GradientOps.cpp:48]: Small but these should probably be moved to each respective heading below.

File: setup.py [hunk 7]
```diff
@@ -28,7 +28,7 @@
     jax_version = [line[4:].strip() for line in f.readlines() if "jax=" in line][0]
 
 requirements = [
-    "pennylane>=0.28",
+    "pennylane @ git+https://github.com/PennyLaneAI/pennylane/@master",
```
Comment (dime10) [inline setup.py:31]: We need to remember to change to this to the next PennyLane release version

**Review by dime10 — COMMENTED** [review]
Thanks for getting the rest of changes in for this one! :)

File: .github/workflows/check-catalyst.yaml (hunk 1 above)
Comment (erick-xanadu) [reply .github/workflows/check-catalyst.yaml:135]: The build process fails when we actually attempt to build with `ccache`. The alternative is to specify `COMPILER_LAUNCHER= ` to leave it empty. I will change it then.

File: frontend/catalyst/compilation_pipelines.py (hunk 3 above)
Comment (erick-xanadu) [reply frontend/catalyst/compilation_pipelines.py:534]: I can change it back. Originally I changed it to avoid some failing tests which are now resolved with #146 . But I think `ResourceWarning` can also be used to indicate that it is taking more computational resources than expected (I.e., we are recompiling therefore wasting resources). Not a big deal to change back.

File: setup.py (hunk 7 above)
Comment (erick-xanadu) [reply setup.py:31]: Yes.

File: frontend/catalyst/jax_primitives.py [hunk 8]
```diff
@@ -1476,3 +1474,10 @@ def _qfor_lowering(
 mlir.register_lowering(func_p, _func_lowering)
 mlir.register_lowering(jvp_p, _jvp_lowering)
 mlir.register_lowering(vjp_p, _vjp_lowering)
+
+# pylint: disable=protected-access
+api_util._shaped_abstractify_handlers[type] = lambda x: core.ShapedArray([], x)
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py:1479]: Oops I think I forgot to push a bunch changes before handing it over, for example here I had the following which might be a bit safer and also creates a weak type:
```py
def _scalar_abstractify(t):
    if isinstance(t, (int, float, complex)):
        raise TypeError(f"Cannot convert given type {t} scalar ShapedArray.")
    return jax.ShapedArray([], dtype=t, weak_type=True)
_shaped_abstractify_handlers[type] = _scalar_abstractify
```

File: .github/workflows/check-catalyst.yaml (hunk 1 above)
Comment (dime10) [reply .github/workflows/check-catalyst.yaml:135]: Yes I think we already use the `COKMPILER_LAUNCHER=""` pattern in several places in the action for exactly this purpose.

File: mlir/Makefile (hunk 5 above)
Comment (dime10) [reply mlir/Makefile:52]: >One is for this https://github.com/tensorflow/mlir-hlo/issues/68 issue. There's an error in mlir-hlo's build system where a dependency is not stated. The error would only show up under certain situations for instance, building in systems with only one processor that forces CMake to execute in a certain order. It doesn't show up when compiling in parallel with multiple processors reliably. Since we compile in github actions, with limited number of processors, the build process fails.

Oh nice good find. We should add a TODO comment here to remove the line once the issue is resolved.

>The other is for removing some MLIR python binding generated code that is repeated in the JAX distribution. During the loading of the module, python complains that these definitions were already loaded.

Got it 👍

File: setup.py (hunk 7 above)
Comment (dime10) [reply setup.py:31]: Another for a TODO comment :)

File: .github/workflows/check-catalyst.yaml (hunk 1 above)
Comment (erick-xanadu) [reply .github/workflows/check-catalyst.yaml]: https://github.com/PennyLaneAI/catalyst/pull/143/commits/7f03eb4ef133ac3d3ee2c691192a892d40b4ff65, https://github.com/PennyLaneAI/catalyst/pull/143/commits/592e288d319e17c95b980d3fd8e0e57d529ae237,
https://github.com/PennyLaneAI/catalyst/pull/143/commits/7d02284d887dff83a284f72b7af445d85b52ebeb

File: mlir/Makefile (hunk 5 above)
Comment (erick-xanadu) [reply mlir/Makefile:53]: https://github.com/PennyLaneAI/catalyst/pull/143/commits/7db5fc095a8afb5d1b372e009ef3e2668598b704

File: setup.py (hunk 7 above)
Comment (erick-xanadu) [reply setup.py:32]: https://github.com/PennyLaneAI/catalyst/pull/143/commits/540b379c931696dbd48da6e663790337af264da7

File: mlir/lib/Gradient/IR/GradientOps.cpp (hunk 6 above)
Comment (erick-xanadu) [reply mlir/lib/Gradient/IR/GradientOps.cpp]: https://github.com/PennyLaneAI/catalyst/pull/143/commits/3f9283ed6c9bac1fed84e1dd901e980bae278ef0

File: frontend/catalyst/jax_primitives.py (hunk 4 above)
Comment (dime10) [reply frontend/catalyst/jax_primitives.py:23]: If this is too much work you can also skip this comment and we can look at our usage of the JAX API in the future.

**Comment by dime10** [comment]
@erick-xanadu thanks for pointing this out, we'll indeed have to make some adjustments!

File: frontend/catalyst/jax_primitives.py (hunk 8 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py]: https://github.com/PennyLaneAI/catalyst/pull/143/commits/e63b067523b92e09239a8f0ff2c769a5c1dd4b41

File: frontend/catalyst/jax_tracer.py [hunk 9]
```diff
@@ -525,7 +499,7 @@ def custom_lower_jaxpr_to_module(
         # register_dialect()
         # Remove module name characters that XLA would alter. This ensures that
         # XLA computation preserves the module name.
-        module_name = _module_name_regex.sub("_", module_name)
+        # module_name = _module_name_regex.sub("_", module_name)
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/jax_tracer.py]: https://github.com/PennyLaneAI/catalyst/pull/143/commits/91fab235eb1f05020f19d7501b48e453887c385d

File: frontend/catalyst/jax_primitives.py (hunk 4 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:24]: Let's just leave that for another ticket, but I see exactly what you mean. Thanks for pointing it out!

**Review by dime10 — APPROVED** [review]
Great stuff 🎉

File: doc/changelog.md [hunk 10]
```diff
@@ -145,6 +145,9 @@
 * Use Tablegen to define MLIR passes instead of C++ to reduce overhead of adding new passes.
   [#157](https://github.com/PennyLaneAI/catalyst/pull/157)
 
+* Add support for JAX v4.10
```
Comment (dime10) [inline doc/changelog.md:148]: "Adding support" for JAX `v0.4.10` is not the right phrasing, since we are switching our (hard) dependency chain to this version rather than adding support for it.

Comment (erick-xanadu) [reply doc/changelog.md]: Thanks! I see what you mean, what do you think about the new comment here:

> Update JAX to `v.0.4.10`.

https://github.com/PennyLaneAI/catalyst/pull/143/commits/3b9ab5e7c132214034304272bcc0f8ec9a149a5a

Comment (dime10) [reply doc/changelog.md]: Thanks!
